base_local_planner/src/critics/speed_cost_function.cpp

```diff
@@ -0,0 +1,219 @@
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of Willow Garage, Inc. nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Daniel Grieneisen
+ *********************************************************************/
+
+#include <base_local_planner/critics/speed_cost_function.h>
+#include <tf/transform_datatypes.h>
+
+namespace base_local_planner {
+
+SpeedCostFunction::SpeedCostFunction() :
+    max_linear_velocity_(1.0),
+    min_linear_velocity_(0.0),
+    max_allowed_vel_(1.0),
+    linear_acceleration_(0.7),
+    half_angle_(1.6),
+    min_distance_to_stop_(0.0),
+    max_distance_to_stop_(0.0),
+    y_buffer_(0.3),
+    x_buffer_(0.3),
+    footprint_min_y_(-0.4),
+    footprint_min_x_(-0.4),
+    footprint_max_y_(0.4),
+    footprint_max_x_(0.4),
+    body_frame_id_("base_link"),
+    world_frame_id_("odom") {}
+
+bool SpeedCostFunction::prepare() {
+  // Reset the maximum allowed velocity
+  max_allowed_vel_ = max_linear_velocity_;
+  if (!obstructions_)
+  {
+    ROS_WARN_THROTTLE(1.0, "No obstructions");
+    return true;
+  }
+
+  // Find the nearest obstruction to the robot that is within the allowed range
+  // Loop over all of the obstructions
+  for (auto obs : (*obstructions_))
+  {
+    // Skip non-dynamic things or things that have been cleared
+    if (obs.type != costmap_2d::ObstructionMsg::DYNAMIC || obs.cleared)
+    {
+      continue;
+    }
+    costmap_2d::ObstructionMsg obs_body_frame = obstructionToBodyFrame(obs);
+    // Check if the bearing to the obstacle is acceptable
+    if (std::fabs(getBearingToObstacle(obs_body_frame)) > half_angle_)
+    {
+      continue;
+    }
+
+    double speed = calculateAllowedSpeed(obs_body_frame);
+    if (speed < max_allowed_vel_)
+    {
+      max_allowed_vel_ = speed;
+    }
+  }
+  ROS_INFO_THROTTLE(0.2, "Setting max speed to %f", max_allowed_vel_);
```

Comment:
Should probably move this into the if that sets the speed, not sure how spammy this is but maybe should be a DEBUG?

Reply:
Dpme